### PR TITLE
Macro Environment

### PR DESCRIPTION
**Pull request by Robadob**
Work in progress support for large environment properties with atomic access from agent functions.

**Progress**
- [x] `EnvironmentDescription` changes
- [x] New `CUDAMacroEnvironment` class
- [x] `CUDASimulation` integration
- [x] `DeviceAPI` access
- [x] Seatbelts support for preventing atomic/read conflicts
- [x] `HostAPI` access
- [x] Submodel mapping support
- [x] Tests
- [x] RTC support
- [x] SWIG support
- [x] SWIG Tests
- [x] docs.fgpu (https://github.com/FLAMEGPU/FLAMEGPU2-docs/pull/60)
    - [x] Document clearly that py doesn't support overiding operator=, so have to use `[0]` or `set()`
- [x] ReadOnly API support (agent function conditions currently have read/write access to macro env).

**Still Requiring Thought**
* Allowing users to default init these properties
* Seatbelts is guaranteed to protect against double use across multiple agentfn in the same layer? 
    * *Yes*
* Does seatbelts detect a host fn, trying to access a property during the same layer?
    * *Host functions are not currently allowed in the same layer as agent functions (or other host functions), although before I realised this I added a checking mechanism anyway.*

**Misc**
* Renamed, `AgentEnvironment` tests to `DeviceEnvironment`, this is more consistent with everything else.

Addresses #162, #308

### CODE REVIEW

**Comment by Robadob** [comment]
Code changes should be finished now. Just docu comments (and separate user guide PR) remain.

All python tests pass (Debug + Seatbelts)
image

**Comment by Robadob** [comment]
In writing docs, have noticed a few safety considerations I need to review as I can't recall if they were implemented as fully as possible.

**Comment by Robadob** [comment]
Discussed name change in meeting. Agreed to keep the same. This will mean removing the inheritance between `DeviceAPI` and `ReadOnlyDeviceAPI` to avoid wasting registers by hiding `ReadOnlyDeviceEnvironment` inside regular `DeviceAPI`.

**Comment by Robadob** [comment]
Well I've now split DeviceAPI into mutable/immutable locally (not commit yet), but 5-6 tests are failing in release+seatbelts mode which I need to investigate. (6th being concurrency test)

```
[ RUN      ] DeviceMacroPropertyTest.max2
c:\users\robadob\fgpu2\tests\test_cases\runtime\test_device_macro_property.cu(457): error: Expected: cudaSimulation.simulate() doesn't throw an exception.
  Actual: it throws class flamegpu::exception::DeviceError with description "Device function 'init' reported 1 errors.
First error:
C:/Users/Robadob/fgpu2/include\flamegpu/runtime/utility/DeviceMacroProperty.cuh(191)[0,0,0][0,0,0]:
DeviceMacroProperty read and atomic write operations cannot be mixed in the same layer as this may cause race conditions.
".
[  FAILED  ] DeviceMacroPropertyTest.max2 (3 ms)
[ RUN      ] DeviceMacroPropertyTest.exchange_int32
c:\users\robadob\fgpu2\tests\test_cases\runtime\test_device_macro_property.cu(608): error: Expected: cudaSimulation.simulate() doesn't throw an exception.
  Actual: it throws class flamegpu::exception::DeviceError with description "Device function 'init' reported 1 errors.
First error:
C:/Users/Robadob/fgpu2/include\flamegpu/runtime/utility/DeviceMacroProperty.cuh(191)[0,0,0][0,0,0]:
DeviceMacroProperty read and atomic write operations cannot be mixed in the same layer as this may cause race conditions.
".
[  FAILED  ] DeviceMacroPropertyTest.exchange_int32 (2 ms)
[ RUN      ] DeviceMacroPropertyTest.exchange_uint32
c:\users\robadob\fgpu2\tests\test_cases\runtime\test_device_macro_property.cu(637): error: Expected: cudaSimulation.simulate() doesn't throw an exception.
  Actual: it throws class flamegpu::exception::DeviceError with description "Device function 'init' reported 1 errors.
First error:
C:/Users/Robadob/fgpu2/include\flamegpu/runtime/utility/DeviceMacroProperty.cuh(191)[0,0,0][0,0,0]:
DeviceMacroProperty read and atomic write operations cannot be mixed in the same layer as this may cause race conditions.
".
[  FAILED  ] DeviceMacroPropertyTest.exchange_uint32 (2 ms)
[ RUN      ] HostMacroPropertyTest.ReadTest
c:\users\robadob\fgpu2\tests\test_cases\runtime\test_host_macro_property.cu(76): error: Expected: cudaSimulation.simulate() doesn't throw an exception.
  Actual: it throws class flamegpu::exception::DeviceError with description "Device function 'agentwrite' reported 120 errors.
First error:
C:/Users/Robadob/fgpu2/include\flamegpu/runtime/utility/DeviceMacroProperty.cuh(191)[0,0,0][0,0,0]:
DeviceMacroProperty read and atomic write operations cannot be mixed in the same layer as this may cause race conditions.
".
[  FAILED  ] HostMacroPropertyTest.ReadTest (1 ms)
[ RUN      ] HostMacroPropertyTest.ZeroTest
c:\users\robadob\fgpu2\tests\test_cases\runtime\test_host_macro_property.cu(208): error: Expected: cudaSimulation.simulate() doesn't throw an exception.
  Actual: it throws class flamegpu::exception::DeviceError with description "Device function 'agentwrite' reported 120 errors.
First error:
C:/Users/Robadob/fgpu2/include\flamegpu/runtime/utility/DeviceMacroProperty.cuh(191)[0,0,0][32,0,0]:
DeviceMacroProperty read and atomic write operations cannot be mixed in the same layer as this may cause race conditions.
".
[  FAILED  ] HostMacroPropertyTest.ZeroTest (2 ms)
```

**Comment by Robadob** [comment]
A different mix, are failing with the same message under Debug.

```
[  FAILED  ] DeviceMacroPropertyTest.RTC_add
[  FAILED  ] DeviceMacroPropertyTest.sub
[  FAILED  ] DeviceMacroPropertyTest.sub2
[  FAILED  ] DeviceMacroPropertyTest.postincrement
[  FAILED  ] DeviceMacroPropertyTest.preincrement
[  FAILED  ] DeviceMacroPropertyTest.min
[  FAILED  ] DeviceMacroPropertyTest.min2
[  FAILED  ] DeviceMacroPropertyTest.max
[  FAILED  ] DeviceMacroPropertyTest.max2
```
Unclear where this has come from, after a month on leave, but will investigate.

**Comment by Robadob** [comment]
Appear to have fixed it (Debug all now pass), I wasn't explicitly memsetting device memory allocated for macro properties. So the read/write flag would sometimes begin non-zero.

**Comment by Robadob** [comment]
> * Does seatbelts detect a host fn, trying to access a property during the same layer?
>     * No, there is currently an explicit synchronisation before host functions, but documentation discourages people from writing models that modify in same layer.

I think I will add this exception before I merged the PR, it can always be softened in future if desired.

**Comment by Robadob** [comment]
* I've now added a check for device write ops before host fn, requires a test adding
* I still need to add a check for device read ops before a host write, this is a bit more fiddly
* It appears that host fn writing to macro property before device fn reads back has not been tested, as I can't find where the upload is triggered to breakpoint it to add the read ops check (it will be delayed but this shouldn't really matter).

**Comment by Robadob** [comment]
Well I did all that work, then discovered that `LayerDescription` won't let a host function be in a layer with anything else, which renders it redundant. I'll leave the code in and the tests commented out. But this should be ready to merge after a squash and all tests have been checked.

**Comment by Robadob** [comment]
```
Win, C++, Debug Seatbelts=ON: 945 pass, 0 fail, 38 disabled
Win, C++, Release Seatbelts=ON: 957 pass, 1 fail, 25 disabled (TestCUDASimulationConcurrency.RTCLayerConcurrency failed, poss bc im running py tests in parallel)
Win, C++, Release Seatbelts=OFF: ----
Win, Py, Debug Seatbelts=ON: 543 pass, 0 fail, 4 skipped
Win, Py, Release Seatbelts=ON:  ----
Win, Py, Release Seatbelts=OFF:  ----
```

File: include/flamegpu/runtime/DeviceAPI.cuh
Comment (ptheywood) [inline include/flamegpu/runtime/DeviceAPI.cuh:95]: ~would `getCurrentStep` be a better name? (just a suggestion while viewing the diff)~

Just spotted this was a move not an addition, so probably better to avoid the breaking change.

File: include/flamegpu/runtime/DeviceAPI.cuh
Comment (ptheywood) [inline include/flamegpu/runtime/DeviceAPI.cuh:437]: Should this DTHROW if setabelts is on instead?

File: include/flamegpu/runtime/utility/DeviceMacroProperty.cuh
Comment (ptheywood) [inline include/flamegpu/runtime/utility/DeviceMacroProperty.cuh:92]: I'm not super keen on these overloads. It hides that atomic ops are being used.

File: include/flamegpu/runtime/utility/DeviceMacroProperty.cuh
Comment (ptheywood) [inline include/flamegpu/runtime/utility/DeviceMacroProperty.cuh:146]: Again i'm not keen on how it is not immediately clear that this is atomic without reading the docs, but that might just be an artifact of using CUDA too much.

File: src/flamegpu/model/SubEnvironmentDescription.cpp
Comment (ptheywood) [inline src/flamegpu/model/SubEnvironmentDescription.cpp:76]: Should this mention macro early in the exception message? (as in exceptions below?)

File: swig/python/flamegpu.i
Comment (ptheywood) [inline swig/python/flamegpu.i:15]: Comment would be good. is this `operator--`?

File: swig/python/flamegpu.i
Comment (ptheywood) [inline swig/python/flamegpu.i:530]: If these commented out ignore/renames are not required could they be removed? Or atleast explained why they are still here?

**Review by ptheywood — CHANGES_REQUESTED** [review]
I'm still not convinced by macro property as a name, but still don't have a better solution. 

Otherwise just some fairly trivial changes to exception messages would be good. Might be worth considering seatbelts vs non-seatbelts performance at some point too, as I expect seatbelts to suffer here. 

CUDA C++ tests all pass under linux with CUDA 11.2 and GCC 9. Python tests currently running but looking promising so far (80%).

File: include/flamegpu/gpu/CUDAMacroEnvironment.h
Comment (Robadob) [reply include/flamegpu/gpu/CUDAMacroEnvironment.h:185]: probably

File: include/flamegpu/gpu/CUDAMacroEnvironment.h
Comment (Robadob) [reply include/flamegpu/gpu/CUDAMacroEnvironment.h:27]: Yeah vs has a thing where it introduces phantom spaces, so i'm never sure if the spaces are real or not.

File: include/flamegpu/runtime/DeviceAPI.cuh
Comment (Robadob) [reply include/flamegpu/runtime/DeviceAPI.cuh:437]: Yes, but that's a separate issue affecting whole of DeviceAPI. Will create a separate issue/pr.

File: include/flamegpu/runtime/utility/DeviceMacroProperty.cuh
Comment (Robadob) [reply include/flamegpu/runtime/utility/DeviceMacroProperty.cuh:146]: The understanding should be that anything except reading the value is atomic in an agent fn with macro env prop.

File: swig/python/flamegpu.i
Comment (Robadob) [reply swig/python/flamegpu.i:15]: That rings a bell, but I can't confirm at this point.

File: swig/python/flamegpu.i
Comment (Robadob) [reply swig/python/flamegpu.i:530]: I'll remove, was probably dev junk from trying to get it to work

**Pull request merged**
